bug-unusable-package

schmelz21 opened the pull request:
### Bug Fix for Issue #5 Package is not Usable

Including the work addressing the expected behavior of `import pe_reports`, additional dev-standard updates need to be made. This include the issues related to the this [LGMT report](https://lgtm.com/projects/g/cisagov/pe-reports/alerts/?mode=list), relative import adjustments and the adding back in additional test from the dev-guidlines skeletons. Detailed notes can be traced back to the previous pull-request [first-commit-V2](https://github.com/cisagov/pe-reports/pull/2).

🧪 **Testing**
Unittests, pytest and pre-commit will be made in addition to steps taken to insure proper installation and usage for `import pe_reports` as a usable package.

✅ **Checklist**
 
- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - eschew scope creep!
- [x] All future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the CONTRIBUTING document.
- [x] These code changes follow cisagov code standards.
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.

schmelz21 committed:
Updated relative imports

schmelz21 committed:
Update setup packages

schmelz21 committed:
Minor code update to pages

schmelz21 committed:
lgtm adjustments

schmelz21 committed:
Removed repeated code

schmelz21 committed:
Update tests

schmelz21 committed:
Added minimal required tests

schmelz21 committed:
linting issue

schmelz21 committed:
Updated TODOs

schmelz21 committed:
Updated styles

schmelz21 commented:
Please review PR "Bug Fix for Issue #5 - Package is not Usable". 
Note the primary issue of not being able to `import pe-report ` with no errors was resolved and tested as follows:
```
(pe-reports) schmelzs@MAC-0019615 pe-reports % python3
Python 3.9.4 (default, Apr 28 2021, 14:10:42) 
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pe_reports
>>> 
```
Secondary efforts included: No LGTM issues, adjusted relative import issues, code improvements and the additional effort to add back in dev-guidline skeleton tests (test_stdout_version, test_running_as_module) successfully. Any open issues were noted in [first-commit-V2](https://github.com/cisagov/pe-reports/pull/2) accordingly.

Note: To introduce back the skeleton tests (test_stdout_version, test_running_as_module), code updates were made in pe_reports.report_generator which include arguments via Docopt module that resulted in the addition of TODOs: "Build code to connect customer db, encrypt and embed pdf reports." and "Add generate_reports func to handle cmd line arguments and function." 

Regardless these items were planned for next pull-request along with a full pe-report buildout once we have this PR completed with the goal of an architecture compliant to the python-skeleton.

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -1,14 +1,30 @@
+"""A tool for creating Posture & Exposure reports.
+
+Usage:
+    pe-reports REPORT_DATE  DATA_DIRECTORY OUTPUT_DIRECTORY
```
Remove unnecessary extra space character
```suggestion
    pe-reports REPORT_DATE DATA_DIRECTORY OUTPUT_DIRECTORY
```

schmelz21 replied:
Resolved in 1299aba

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -1,14 +1,30 @@
+
+Options:
+    -h --help                       Show this message.
+    REPORT_DATE                     Date of the report, format YYYY-MM-DD
```
Add period to be consistent with other lines
```suggestion
    REPORT_DATE                     Date of the report, format YYYY-MM-DD.
```

schmelz21 replied:
Resolved in 1299aba

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -22,12 +38,28 @@ def load_template():
 
+def generate_reports(data, data_dir, out_dir):
+    """Gather assets to produce reports."""
+    # TODO: build code to connect customer db, encrypt and embed pdf reports.
```
Please [create an issue](https://github.com/cisagov/pe-reports/issues/new?assignees=&labels=&template=issue.md) for this TODO item and update this comment to include a link to that issue.

schmelz21 replied:
Issue created: https://github.com/cisagov/pe-reports/issues/7
Noted in commit: 7ba36fa

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -22,12 +38,28 @@ def load_template():
+    """Set up logging and call the pe_reprots function."""
+    args: Dict[str, str] = docopt.docopt(__doc__, version=__version__)
+
+    # TODO: Add generate_reports func to handle cmd line arguments and function.
```
Please [create an issue](https://github.com/cisagov/pe-reports/issues/new?assignees=&labels=&template=issue.md) for this TODO item and update this comment to include a link to that issue.

schmelz21 replied:
Issue created: https://github.com/cisagov/pe-reports/issues/8
Noted in commit: 7ba36fa

dav3r commented on tests/test_pe_reports.py:
```diff
@@ -1,11 +1,61 @@
 # TODO: Replace current dummy test with useful tests - https://github.com/cisagov/pe-reports/issues/3#issue-909531010
 
 
+def test_output(capsys):
+    """Routine to test captured output."""
+    print("hello")
+    sys.stderr.write("world\n")
+    captured = capsys.readouterr()
+    assert captured.out == "hello\n"
+    assert captured.err == "world\n"
+    print("next")
+    captured = capsys.readouterr()
+    assert captured.out == "next\n"
```
This test doesn't test anything in this project- it can be deleted.  The same goes for the `TestingStringMethods` class below.

schmelz21 replied:
Resolved in 124615b

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -14,27 +15,35 @@
+            )
+            font = run.font
+            Paragraph.text_style_title(prs, font)
+        except Exception:
```
We should catch something more specific than `Exception` here and in all similar cases below.

schmelz21 replied:
Resolved in 675a8e4
added in "FileNotFoundError" for the exception.

dav3r reviewed (changes requested):
Looking much better here @schmelz-ctr, but I noted some things I'd like to see cleaned up before I approve.

schmelz21 committed:
Minor text adjustments.

schmelz21 committed:
Updated TODOs Issues references

schmelz21 committed:
Removed unusable tests

schmelz21 committed:
Updated to FileNotFoundError

mcdonnnj commented on src/pe_reports/__init__.py:
```diff
@@ -3,6 +3,7 @@
+
+from ._version import __version__  # noqa: F401
+
+__all__ = ["pages", "report_generator", "stylesheet"]
```
These will only be imported when doing `from pe_reports import *`. I just want to make sure that's the intended functionality. With the `__init__.py` as it is there is nothing directly accessible when doing `import pe_reports`:

```console
>>> import pe_reports
>>> dir(pe_reports)
['__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', '_version']
```

schmelz21 replied:
@mcdonnnj, I understand the issue `import *`.  The change was made initially due to questions brought up in ["first-commit"](https://github.com/cisagov/pe-reports/pull/2#discussion_r645875140), and I assumed it was a skeleton dependancy. For the sake of using `__all__` at this time, I cannot provide a reasonable purpose on my side and would look to remove it, if it makes sense to.

mcdonnnj replied:
The `__all__` member is a Python packaging item. You can see more information about it [in the documentation](https://docs.python.org/3/tutorial/modules.html#importing-from-a-package). If you do not want/need any subpackages to be available for import unless directly imported feel free to remove the line.

schmelz21 replied:
Resolved in ed8f449

mcdonnnj replied:
This appears to have been removed in ed8f449 and then re-added in 016ddbf.

schmelz21 replied:
Sorry, yes bad notation my part. It was re-added so to properly be included as a dependancy at setup. I should have been re-commented "Resolved in 016ddbf"

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -1,14 +1,30 @@
+                                    files are located. Organized by
+                                    owner.
+    OUTPUT_DIRECTORY                The directory where the final PDF
+                                    reports should be saved.
```
```suggestion
Arguments:
  REPORT_DATE       Date of the report, format YYYY-MM-DD.
  DATA_DIRECTORY    The directory where the excel data files are located.
                    Organized by owner.
  OUTPUT_DIRECTORY  The directory where the final PDF reports should be saved.

Options:
  -h --help     Show this message.
  -v --version  Show version information.
```

schmelz21 replied:
Resolved in 74a60f3

mcdonnnj reviewed (changes requested):
This is great progress toward getting this package into a stable-for-development state. I had a question and a suggestion below. However, I would like to reiterate that hard-coding file locations as you have done throughout this package is not viable. The way you have done this requires one to download the entire repository, install from that, and then only run it from within that directory structure. Example:

```console
$ pip install https://github.com/cisagov/pe-reports/archive/bug-unusable-package.zip
```

and then

```console
$ pe-reports $(date "+%Y-%m-%d") data/ output/

 [Info] Loading Posture & Exposure Report Template, Version: 0.0.1
Traceback (most recent call last):
  <traceback information omitted>
  /pptx/opc/phys_pkg.py", line 32, in __new__
    raise PackageNotFoundError("Package not found at '%s'" % pkg_file)
pptx.exc.PackageNotFoundError: Package not found at 'src/pe_reports/data/shell/pe_shell.pptx'
```

Please see [`pkg_resources`](https://setuptools.readthedocs.io/en/latest/pkg_resources.html) as an option to access package data (since we already require `setuptools`):

```console
>>> pkg_resources.resource_filename("pe_reports", "shell/pe_shell.pptx")
'/path/to/venv/lib/python3.9/site-packages/pe_reports/shell/pe_shell.pptx'
```

schmelz21 committed:
Update src/pe_reports/report_generator.py

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

schmelz21 committed:
Update package resource

schmelz21 committed:
Update to package resources

schmelz21 committed:
isort adjustment

schmelz21 commented:
@mcdonnnj, Please re-review.
I have updated the package resources so that you get the following behavior after install.
```
Finished processing dependencies for pe-reports==0.0.1
(pe-reports) schmelzs@MAC-0019615 pe-reports % pe-reports $(date "+%Y-%m-%d") data/ output/                                                  
 [Info] Loading Posture & Exposure Report Template, Version: 0.0.1

 [Info] Generating Graphs 
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_customer.csv' There is no customer data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_creds.csv' There is no credentials data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_domains.csv' There is no domain data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_malware.csv' There is no malware data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_vulns.csv' There is no vulnerability data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_web.csv' There is no web data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_tld_df.csv' There is no tld chart data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_ce_df.csv' There is no ce_df chart data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_web_df.csv' There is no web_df chart data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_ma_df.csv' There is no ma_df chart data.
[Errno 2] No such file or directory: 'src/pe_reports/data/csv/dhs_iv_df.csv' There is no dhs_iv_df chart data.
No output available.
(pe-reports) schmelzs@MAC-0019615 pe-reports % 
```
In addition, I have updated Issue: "Add command line functionality." #8 to include acceptance criteria that addresses the management of input/output files accordingly. These issues are planned for next pull-request along with a full pe-report buildout once we are stable.

schmelz21 commented:
Good Morning, @mcdonnnj and @dav3r is it possible to get a re-review on schedule. Thanks

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -14,27 +15,35 @@
-        font = run.font
-        font = Paragraph.text_style_title(prs, font)
+        try:
+            df_customer = pd.read_csv("src/pe_reports/data/csv/dhs_customer.csv")
```
```suggestion
            # TODO: Remove hard-coded file locations
            # See https://github.com/cisagov/pe-reports/issues/8
            df_customer = pd.read_csv("src/pe_reports/data/csv/dhs_customer.csv")
```

jsf9k replied:
I'd also strongly suggest including such a comment every time a hard-coded file is referenced below.

schmelz21 replied:
Resolved in 4c928bd

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -22,12 +39,28 @@ def load_template():
+    try:
+        pptx_out = "Customer_ID_Posture_Exposure.pptx"
+        prs.save(os.path.join("/output", pptx_out))
+    except OSError:
+        print("No output available.")
```
Can this be modified so that the exception output is also printed here?  `No output available` seems a bit vague to me.

schmelz21 replied:
Resolved in 4c928bd

jsf9k replied:
I don't think commit 4c928bd added the exception output to this particular case.

schmelz21 replied:
@jsf9k yes that was a mistake. It is now resolved with commit - 882be6a

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -22,12 +39,28 @@ def load_template():
+
+
 def main():
+    """Set up logging and call the pe_reprots function."""
```
Logging is not being set up here, though it should be.  For reference:
https://github.com/cisagov/skeleton-python-library/blob/4ef2364701c0db074c423f740062b2d7c197bea8/src/example/example.py#L83-L86

schmelz21 replied:
Resolved in 4c928bd

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -22,12 +39,28 @@ def load_template():
+
+
 def main():
+    """Set up logging and call the pe_reprots function."""
```
There is no function called `pe_reprots` in here:
```suggestion
    """Set up logging and call the generate_reports function."""
```

schmelz21 replied:
Resolved in fb98057

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -22,12 +39,28 @@ def load_template():
+    )
+
     """Generate PDF reports."""
     print("\n [Info] Loading Posture & Exposure Report Template, Version:", __version__)
```
Once you set up logging correctly, all `print` statements should be replaced with `logging` statements.

schmelz21 replied:
Resolved in 4c928bd

jsf9k commented on src/pe_reports/pages.py:
```diff
@@ -44,107 +53,134 @@ def overview(prs):
-
-        chart.categories = list(tld_df.columns.values)
-        chart.add_series("Top Level Domains", list(tld_df.loc[0]))
-        x, y, cx, cy = Inches(3.5), Inches(2.9), Inches(2), Inches(1.5)
```
The hard-coded values that appear here and later in the code seem fragile.  Can they be computed somehow, or at a minimum stored in constant variables at the top of the file?

schmelz21 replied:
@jsf9k the `x, y, cx, cy` list as follows:

- Inches(3.5), Inches(2.9), Inches(2), Inches(1.5)
- Inches(5.75), Inches(2.9), Inches(2), Inches(1.5)
- Inches(3.5), Inches(5), Inches(4.8), Inches(2.15)
- Inches(8.25), Inches(2.9), Inches(4.5), Inches(2.0)
- Inches(8.25), Inches(4.9), Inches(4.5), Inches(2.0)

Each has a slight variance to the inputs currently. Also this portion may be expanded or changed on next iteration when a full pe-report build occurs. At that time if we discover patterns in the graph positioning, we can look at constants. Looking to let this one pass for now on the PR.

schmelz21 replied:
Commit referring to hard coded file-paths resolved in 4c928bd

jsf9k replied:
@schmelz-ctr, can you create an issue to remind us to revisit the hard-coded `x`, `y`, `cx`, and `cy` later?  I'll resolve this conversation once (1) the issue is created and (2) a comment referencing the issue is added to this file at each location where these values are defined.

jsf9k replied:
@schmelz-ctr created #9 in response to this comment.

dav3r reviewed (changes requested):
There are still some things that I would like to see fixed here.

jsf9k commented on src/pe_reports/stylesheet.py:
```diff
@@ -69,32 +69,22 @@ def text_style_ov_val(self, font):
-        chart.font.rgb = (20, 200, 100)
+    def ch_font_attr(self, chart):
+        """Font attr set."""
+        self.sm = chart.font.size = Pt(10)
+        self.lg = chart.font.size = Pt(28)
+        self.blue = chart.font.rgb = (20, 200, 100)
+        self.white = chart.font.rgb = (255, 255, 255)
```
These values should really be _static_ constants, as they are the same for all object instances.

schmelz21 replied:
@jsf9k I created constants for "Font Attributes" that can be applied across all graphic components using "fonts". This approach can be expanded as we build out a full report with various attributes.

Resolved in 3ff00e0

jsf9k reviewed (commented):
I made a few comments for your consideration.  I think the only one that I'd _have_ to see before approving is the issue references when using a hard-coded path.

schmelz21 committed:
Logging and code notations

schmelz21 committed:
Adjusted constants

schmelz21 committed:
lgtm constant error

schmelz21 committed:
black linting

schmelz21 committed:
setup adjustment

schmelz21 commented:
@dav3r and @jsf9k the PR is ready for a re-review.

jsf9k reviewed (changes requested):
There is one place where you failed to resolve an issue in the spot where @dav3r pointed it out.  Please fix that and then re-request review from me.

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -1,17 +1,38 @@
+                    Organized by owner.
+  OUTPUT_DIRECTORY  The directory where the final PDF reports should be saved.
+
+Options:
+  -h --help     Show this message.
+  -v --version  Show version information.
+  --log-level=LEVEL     If specified, then the log level will be set to
+                        the specified value.  Valid values are "debug", "info",
+                        "warning", "error", and "critical". [default: info]
```
This formatting should be consistent. There should be two spaces between the longest option and its description and every other option should be spaced to match.

```suggestion
Options:
  -h --help          Show this message.
  -v --version       Show version information.
  --log-level=LEVEL  If specified, then the log level will be set to
                     the specified value.  Valid values are "debug", "info",
                     "warning", "error", and "critical". [default: info]
```

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -22,21 +43,49 @@ def load_template():
+    """Set up logging and call the generate_reports function."""
+    args: Dict[str, str] = docopt.docopt(__doc__, version=__version__)
+
+    # Set up logging
+    logging.basicConfig(
+        format="%(asctime)-15s %(levelname)s %(message)s", level=logging.INFO
+    )
```
Since you are now taking the logging level as a command line option you need to use it in your logging setup.

```suggestion
    # Set up logging
    logging.basicConfig(
        format="%(asctime)-15s %(levelname)s %(message)s",
        level=args["--log-level"].upper(),
    )
```

schmelz-ctr replied:
Resolved in 253b2dd

schmelz21 replied:
Resolved in ecea2ce

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -22,21 +43,49 @@ def load_template():
+
 
 if __name__ == "__main__":
     sys.exit(main())
```
You do not return anything from `main()`. This should only be included if the function has return values indicating normal behavior or errors.

```suggestion
    main()
```

dav3r replied:
Now that `main()` returns a 0 or a 1, you should put this back the way it was:
```
sys.exit(main())
```

schmelz21 replied:
Resolved in 7a0d456

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -1,17 +1,38 @@
 # Standard Python Libraries
+import logging
 import os
 import sys
```
With the removal at the bottom of the file, this import is no longer used.

schmelz21 replied:
@mcdonnnj, I am not sure I follow. If removed, I begin to get import errors. For now I left "import logging" in so that the code passes pre-commit.

mcdonnnj replied:
@schmelz-ctr The line I am providing feedback on is `import sys`. That is why you are now failing checks.

schmelz21 replied:
Now included with the addition of logging test. 253b2dd

mcdonnnj commented on tests/test_pe_reports.py:
```diff
@@ -1,33 +1,43 @@
-        self.assertTrue("TUTORIALSPOINT".isupper())
-        self.assertFalse("TUTORIALSpoint".isupper())
+# TODO: Replace current dummy test with useful tests - https://github.com/cisagov/pe-reports/issues/3#issue-909531010
 
```
Since you have added the `--log-level` option to the main script you should add the [appropriate test from cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library/blob/4ef2364701c0db074c423f740062b2d7c197bea8/tests/test_example.py#L74-L86).

schmelz21 replied:
Resolved in 253b2dd

schmelz-ctr replied:
Resolved in 253b2dd

jsf9k replied:
You should include [this test](https://github.com/cisagov/skeleton-python-library/blob/4ef2364701c0db074c423f740062b2d7c197bea8/tests/test_example.py#L89-L93) as well.

schmelz21 replied:
Added 'bad_log_level' test - Resolved in 25682d6

mcdonnnj reviewed (changes requested):
I had some additional changes to request to clean up some of the things that have been added.

schmelz21 committed:
corrected missed exception issue

schmelz21 committed:
Update src/pe_reports/report_generator.py

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

schmelz21 committed:
Update src/pe_reports/report_generator.py

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

schmelz21 committed:
logging test

schmelz21 committed:
updated test

schmelz21 committed:
Removed print statement

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -14,27 +18,37 @@
+            font = run.font
+            Paragraph.text_style_title(prs, font)
+        except FileNotFoundError as not_found:
+            logging.info(f"{not_found} : There is no customer data.")
```
This and all similar lines to log an exception should use `logging.error` or possibly `logging.exception` if you want to output additional traceback info:
```suggestion
            logging.error(f"{not_found} : There is no customer data.")
```

schmelz21 replied:
Resolved in a8ccd82

schmelz21 committed:
change info to exceptions error

schmelz21 commented:
Quick note - The missed exception issue in the first review was resolved in 882be6a. 
@mcdonnnj , @dav3r , @jsf9k submitting for a re-review. Thanks.

schmelz21 committed:
reset main exit

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -22,21 +44,73 @@ def load_template():
+
     """Generate PDF reports."""
-    print("\n [Info] Loading Posture & Exposure Report Template, Version:", __version__)
+    logging.info(
+        f"[Info] Loading Posture & Exposure Report Template, Version :, {__version__}"
+    )
```
Now that you are using `logging`, the `[Info]` string in this message (and some other places) is redundant.
```suggestion
    logging.info(
        f"Loading Posture & Exposure Report Template, Version :, {__version__}"
    )
```

schmelz21 replied:
Resolved in 10a9aff

schmelz21 committed:
Looging info cleanup

dav3r reviewed (approved):
I believe all of my requested changes have been made - thanks! 👍 

@mcdonnnj and @jsf9k - please circle back to this PR and confirm that your feedback has been suitably addressed.

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -14,27 +18,37 @@
+            font = run.font
+            Paragraph.text_style_title(prs, font)
+        except FileNotFoundError as not_found:
+            logging.error(f"{not_found} : There is no customer data.")
```
Due to particulars of how Python's `logging` library is designed you should use the `printf` style for logging statements. This should be applied to all of the `logging` statements that have variables used in their output.

```suggestion
            logging.error("%s : There is no customer data.", not_found)
```

schmelz21 replied:
Resolved in 4d64c0f.

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -22,21 +44,73 @@ def load_template():
+        pptx_out = "Customer_ID_Posture_Exposure.pptx"
+        prs.save(os.path.join("/output", pptx_out))
+    except FileNotFoundError as not_found:
+        logging.error(f"{not_found} : Missing input data. No report generated.")
```
Mentioning the same thing in this file. Additionally for this line is the error message accurate?

```suggestion
        logging.error("%s : Missing input data. No report generated.", not_found)
```

schmelz21 replied:
Resolved in 4d64c0f

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -22,21 +44,73 @@ def load_template():
+
+    # TODO: Add generate_reports func to handle cmd line arguments and function.
+    # Issue #8: https://github.com/cisagov/pe-reports/issues/8
+    generate_reports(
+        args["REPORT_DATE"], args["DATA_DIRECTORY"], args["OUTPUT_DIRECTORY"]
+    )
```
```suggestion
    generate_reports(
        validated_args["REPORT_DATE"],
        validated_args["DATA_DIRECTORY"],
        validated_args["OUTPUT_DIRECTORY"],
    )
```

schmelz21 replied:
Resolved in 4d64c0f

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -22,21 +44,73 @@ def load_template():
-    prs.save(os.path.join(SAVE_PATH, pptx_out))
+    try:
+        pptx_out = "Customer_ID_Posture_Exposure.pptx"
+        prs.save(os.path.join("/output", pptx_out))
```
Why aren't you using the `OUTPUT_DIRECTORY` argument here?

schmelz21 replied:
Resolved in 4d64c0f

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -22,21 +44,73 @@ def load_template():
-    pptx_out = "Customer_ID_Posture_Exposure.pptx"
-    prs.save(os.path.join(SAVE_PATH, pptx_out))
+    try:
+        pptx_out = "Customer_ID_Posture_Exposure.pptx"
```
This should be declared as a global like with `REPORT_SHELL`.

schmelz21 replied:
Resolved in 4d64c0f

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -22,21 +44,73 @@ def load_template():
     prs = load_template()
 
-    print("\n [Info] Generating Graphs ")
+    logging.info("Generating Graphs ")
```
```suggestion
    logging.info("Generating Graphs")
```

schmelz21 replied:
Resolved in 4d64c0f

mcdonnnj reviewed (changes requested):
Thanks for your continued work in refining this PR. I had some additional changes I would like to see based on the code as it is now.

jsf9k commented on tests/test_pe_reports.py:
```diff
@@ -1,33 +1,79 @@
+# cisagov Libraries
+import pe_reports.report_generator
+
+log_levels = (
+    "debug",
+    "info",
+    "warning",
+    "error",
+    "critical",
+)
```
Is `log_levels` used anywhere in this file?  If not then it should be removed.

It would be nice to see the log level verified in `test_log_levels()` but that may be a change that's better to make to the skeleton itself.  What say you, @mcdonnnj?

mcdonnnj replied:
It should see usage with
https://github.com/cisagov/pe-reports/blob/10a9aff84e6d2ec21e33fc190524cfc3f6798f2f/tests/test_pe_reports.py#L57
which is then used in
https://github.com/cisagov/pe-reports/blob/10a9aff84e6d2ec21e33fc190524cfc3f6798f2f/tests/test_pe_reports.py#L68

jsf9k replied:
OK, thanks for pointing out that `log_levels` is actually used.  My ancient eyes don't see like they used to!

I still don't see where in `test_log_levels()` we actually verify that the log level has been set.

dav3r replied:
It calls `pe_reports.report_generator.main()` with each (valid) log level and tests [whether the root logger has a handler](https://github.com/cisagov/pe-reports/blob/4d64c0f442b65f47d799e7809bd4a68fa395dcb6/tests/test_pe_reports.py#L75-L78).  This test is virtually unchanged from [our skeleton](https://github.com/cisagov/skeleton-python-library/blob/4ef2364701c0db074c423f740062b2d7c197bea8/tests/test_example.py#L74-L86).

jsf9k replied:
Right, I guess I am saying that the skeleton test needs some additional content.  It doesn't actually check that the correct log level was assigned.  If (being specific to this PR) the function `pe_reports.report_generator.main()` just set the logging level to `DEBUG` no matter what you passed in on the command line, then `test_log_levels()` would still pass.

I will create an issue in [cisagov/skeleton-python-library](https://github.com/cisagov/skeleton-python-library) and we can move the discussion there.

jsf9k replied:
Please continue this discussion at cisagov/skeleton-python-library#79.

dav3r replied:
I understand what you are saying now and agree that would be a good test to add to the skeleton.

jsf9k reviewed (changes requested):
I requested a few minor changes, but nothing that should be difficult or take a lot of time to complete.

schmelz21 committed:
Added test bad_log_level

schmelz21 committed:
Log and var updates

schmelz21 committed:
TODO remove hard-code values

schmelz21 committed:
TODO spelling

schmelz21 commented:
@mcdonnnj, @jsf9k please see the updates to the requested changes.
Also note Issue 'Define constants for graph sizing and positioning' https://github.com/cisagov/pe-reports/issues/9  was created and TODOs created in commit d1871d0.
Thanks

schmelz21 commented:
@mcdonnnj and @jsf9k, are you able to schedule a re-review for today?

jsf9k reviewed (approved):
Many thanks to @schmelz-ctr for addressing the concerns I brought up!

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -20,22 +43,74 @@ def load_template():
+        prs.save(os.path.join(out_dir, REPORT_OUT))
+    except FileNotFoundError as not_found:
+        logging.error("%s : Missing input data. No report generated.", not_found)
     return
```
This does nothing currently. You should remove any empty `return` statements in the project as they are unnecessary.

```suggestion
```

schmelz21 replied:
Resolved in 4088918

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -20,22 +43,74 @@ def load_template():
+
     """Generate PDF reports."""
-    print("\n [Info] Loading Posture & Exposure Report Template, Version:", __version__)
+    logging.info(f"Loading Posture & Exposure Report Template, Version : {__version__}")
```
Do not use f-strings in calls to the `logging` library.

```suggestion
    logging.info("Loading Posture & Exposure Report Template, Version : %s", __version__)
```

schmelz21 replied:
Resolved in 4088918

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -20,22 +43,74 @@ def load_template():
+        validated_args["OUTPUT_DIRECTORY"],
+    )
+
     """Generate PDF reports."""
```
This should either be removed or changed to a comment. This was the previous docstring for the `main()` function.

schmelz21 replied:
Removed in 4088918

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -5,6 +5,13 @@
+WHITE = RGBColor(255, 255, 255)
+
 
 class Paragraph:
```
Should these methods be static? They do not appear to use `self` at all.

schmelz21 replied:
@mcdonnnj  I had some back and forth with @jsf9k on this issue. - https://github.com/cisagov/pe-reports/pull/6/files/faf20b39ebfee1b21a034c622d448c483e4a427f#r651176744
I guess the usage is that the values creating the colors like the font sizes "Small", "Large" will never change and considered a value. Like a css element that aligns with brand standards. Scope may change on a full report buildout, but if it is a pythonic issue, I'd look for any suggestions to resolve.

schmelz-ctr replied:
> shape

@mcdonnnj I am not sure I am seeing the issue clearly? Can you elaborate where the problem is showing.
Thanks...

mcdonnnj replied:
Sorry, what I'm referring to specifically is that these methods do need to be in an object to track state. They are just convenience utility methods to get desired formatting. As an example

```python
def text_style_title(self, font):
        """Text style for cover page title."""
        font.name = "Calibri"
        font.size, font.color = LARGE, WHITE
        return font
```

would instead be written as

```python
@staticmethod
def text_style_title(font):
        """Text style for cover page title."""
        font.name = "Calibri"
        font.size, font.color = LARGE, WHITE
        return font
```

Unless the methods are static you will need to create an object (`paragraph_formatter = Paragraph()`) to use these methods.

schmelz21 replied:
Resolved in f04e778

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -47,54 +54,34 @@ def text_frame_key_metric(self, slide, shape, name):
         chart.has_legend = True
         chart.legend.position = XL_LEGEND_POSITION.BOTTOM
         chart.legend.include_in_layout = False
         return
```
```suggestion
```

schmelz21 replied:
Resolved in b5cf970

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -47,54 +54,34 @@ def text_frame_key_metric(self, slide, shape, name):
         return font
 
 
 class Graph:
```
Same question about static methods.

schmelz21 replied:
Resolved in f04e778

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -47,54 +54,34 @@ def text_frame_key_metric(self, slide, shape, name):
+    def chart_sm(self, slide, chart):
+        """Small chart."""
+        chart.font.size, chart.font.color = SMALL, GREEN
         return
```
```suggestion
```

schmelz21 replied:
Resolved in b5cf970

mcdonnnj reviewed (changes requested):
We're down to some last polish now. Thanks again for your work. In addition to these changes I wanted to mention that in
https://github.com/cisagov/pe-reports/blob/d1871d03112d428b2b95ee57f08061f0b2166222/src/pe_reports/stylesheet.py#L38-L52
the `shape` argument is redefined. This should be fixed.

schmelz21 committed:
guidline format changes

schmelz21 committed:
Stylesheet format changes

schmelz21 commented:
> We're down to some last polish now. Thanks again for your work. In addition to these changes I wanted to mention that in
> https://github.com/cisagov/pe-reports/blob/d1871d03112d428b2b95ee57f08061f0b2166222/src/pe_reports/stylesheet.py#L38-L52
> 
> 
> the `shape` argument is redefined. This should be fixed.

@mcdonnnj - I am unsure of the change on this issue. Can you elaborate?

dav3r commented:
> > We're down to some last polish now. Thanks again for your work. In addition to these changes I wanted to mention that in
> > https://github.com/cisagov/pe-reports/blob/d1871d03112d428b2b95ee57f08061f0b2166222/src/pe_reports/stylesheet.py#L38-L52
> > 
> > the `shape` argument is redefined. This should be fixed.
> 
> @mcdonnnj - I am unsure of the change on this issue. Can you elaborate?

`shape` is an argument passed in to both of those functions, but that value is not ever used because it is immediately overwritten by the `for` loop, so there is no need to pass `shape` into those functions (unless this is unintended behavior).

schmelz21 committed:
Cleanup invalid arguments

schmelz21 commented:
@mcdonnnj, @dav3r after applying `@staticmethod` to the stylesheet.py components and re-auditing all remaining arguments that may have been redundant and/or of non-use, the code is much cleaner. Looking for a re-review.

Thanks,

schmelz21 commented:
@mcdonnnj with the merging of the "Lineage CONFLICT PR", are we able to proceed with your re-review. Thanks.

jsf9k committed:
Merge branch 'develop' into bug-unusable-package

schmelz21 commented:
@mcdonnnj are you able to schedule a review?

dav3r reviewed (approved):
LGTM! 🦖

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -5,96 +5,91 @@
         for shape in slide.shapes:
             if not shape.has_text_frame:
                 continue
         return shape
```
You need to define this variable in case the `slide.shapes` iterable is empty.

schmelz21 replied:
Resolved in d6bab3f

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -5,96 +5,91 @@
         for shape in slide.shapes:
             print(shape.name)
-        return
+        return shape
```
You need to define this variable in case the `slide.shapes` iterable is empty.

schmelz21 replied:
Resolved in d6bab3f

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -14,137 +18,203 @@
+            font = run.font
+            Paragraph.text_style_ov_val(font)
+        except FileNotFoundError as not_found:
+            logging.error("%s : There is no customer data.", not_found)
```
This same pattern appears to be used five times in a row. Please make an issue (and add comments referencing the issue) to DRY out this code.

schmelz21 replied:
Reference added in 36029f4. 
Issue created - https://github.com/cisagov/pe-reports/issues/18
Note: There will be a change in the graphing tools to Matplotlib next iteration. Added as an "Acceptance Criteria" for future dev - https://github.com/cisagov/pe-reports/issues/14

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -14,137 +18,203 @@
+            ).chart
+            Graph.chart_sm(chart)
+        except FileNotFoundError as not_found:
+            logging.error("%s : There is no customer data.", not_found)
```
This is another chunk of code that appears to be repeating a core pattern. Please make an issue (and comments) to DRY out this code.

schmelz21 replied:
Reference added in 36029f4. 
Issue created - https://github.com/cisagov/pe-reports/issues/18
Note: There will be a change in the graphing tools to Matplotlib next iteration. Added as an "Acceptance Criteria" for future dev - https://github.com/cisagov/pe-reports/issues/14

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -20,22 +43,74 @@ def load_template():
+    # TODO: Add generate_reports func to handle cmd line arguments and function.
+    # Issue #8: https://github.com/cisagov/pe-reports/issues/8
+    generate_reports(
+        validated_args["REPORT_DATE"],
+        validated_args["DATA_DIRECTORY"],
+        validated_args["OUTPUT_DIRECTORY"],
```
Please make an issue (and add a comment referencing that issue) to improve the use of `schema` to actually validate these arguments. You give an expected format for `REPORT_DATE` and `DATA_DIRECTORY` should exist at a minimum.

schmelz21 replied:
Reference added in 36029f4. 
Issue created - https://github.com/cisagov/pe-reports/issues/19

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -5,96 +5,91 @@
             if shape.name == name:
                 text_frame = shape.text_frame
                 text_frame.clear()
         return text_frame
```
You need to define this variable in case there is no `shape` in `slide.shapes` such that `shape.name == name`.

schmelz21 replied:
Resolved in d6bab3f

mcdonnnj replied:
It looks like this one got missed for adding a comment referencing #21.

mcdonnnj commented on src/pe_reports/stylesheet.py:
```diff
@@ -5,96 +5,91 @@
             if shape.name == name:
                 text_frame = shape.text_frame
                 text_frame.clear()
         return text_frame
```
You need to define this variable in case there is no `shape` in `slide.shapes` such that `shape.name == name`.

schmelz21 replied:
@mcdonnnj, a question on approach. The python-pptx package natively builds out a 'shape' object, as continuation or layer. (pptx file--> slide--> shape(ex. text box)). From [pptx-python](https://python-pptx.readthedocs.io/en/latest/user/understanding-shapes.html)

```
`from pptx import Presentation

SLD_LAYOUT_TITLE_AND_CONTENT = 1

prs = Presentation()
slide_layout = prs.slide_layouts[SLD_LAYOUT_TITLE_AND_CONTENT]
slide = prs.slides.add_slide(slide_layout)
shapes = slide.shapes

for shape in slide.shapes:
    print("out : ", shape)
```
`out :  <pptx.shapes.placeholder.SlidePlaceholder object at 0x7fa7f80658b0>`

To declare a variable that passes on an empty iterable, I am assuming I need to create a "default-shape-object" to pass code freely between functions/models and pass tests. With that said we have sunsetted the python-pptx graphing tools for Matplotlib graphing tools next iteration. https://github.com/cisagov/pe-reports/issues/14. 

Instead of making additional code changes here, is it acceptable to create an Issue for future development when adding in the Matplotlib graphs- "Define variable for empty interation"? This would then be an added criteria to https://github.com/cisagov/pe-reports/issues/14.

The purpose of this merge would then satisfy the compliance to the [skeleton-python-library](https://github.com/cisagov/skeleton-python-library).

mcdonnnj replied:
Yes creating an issue and tying it to the work in #14 is reasonable.

schmelz21 replied:
Resolved in d6bab3f

mcdonnnj reviewed (changes requested):
Thank you for your continued work on getting this finished. Please see some comments I had about adding issues for some code cleanup/improvement as well as covering potential fail states. We are on the home stretch at this point I believe.

schmelz21 committed:
Add Issue refereces

schmelz21 committed:
Added issue empty iterable

schmelz21 commented:
@mcdonnnj thanks for the quick response. The [issue](https://github.com/cisagov/pe-reports/issues/21) has been made and referenced accordingly. Requesting a re-review with a merge to 'develop'. - Thanks

mcdonnnj commented on src/pe_reports/pages.py:
```diff
@@ -14,137 +18,233 @@
+            df_creds = pd.read_csv("src/pe_reports/data/csv/dhs_creds.csv")
+
+            # TODO: Dry out this code
+            # Issue 8: https://github.com/cisagov/pe-reports/issues/18
```
```suggestion
            # Issue 18: https://github.com/cisagov/pe-reports/issues/18
```

schmelz21 replied:
Thats my miss. Taking care of it now.

schmelz21 replied:
Resolved in 86feeaf.

mcdonnnj reviewed (changes requested):
@schmelz-ctr Just some last documentation clean-up and then this should be in a merge-able state. Thanks for your work on this PR!

schmelz21 committed:
Issues refereces

schmelz21 commented:
@mcdonnnj just resolve those last Issue references - 86feeaf. 
Thanks

mcdonnnj reviewed (approved):
LGTM ✔ Thanks again for your continued work on this PR!
